src/KingSystem/World/worldShootingStarMgr.cpp

```diff
@@ -1 +1,164 @@
+#include "KingSystem/GameData/gdtManager.h"
+#include "KingSystem/World/worldManager.h"
+
+namespace ksys::world {
+
+static bool sScheduled = true;
+
+static struct {
+    int end = 3;
+    int start = 21;
+    int unk = 4;
+} sHours;
+
+static sead::Vector3f sStarProperty{};
+
+ShootingStarMgr::ShootingStarMgr() = default;
+
+ShootingStarMgr::~ShootingStarMgr() = default;
+
+void ShootingStarMgr::init_(sead::Heap* heap) {
+    if (Manager::instance()->isDemo())
+        sScheduled = false;
+    else
+        initSchedule();
+}
+
+// NON_MATCHING
+void ShootingStarMgr::initSchedule() {
+    if (sHours.start >= sHours.end) {
+        sead::FixedObjArray<u32, 24> validHours;
+
+        if (sHours.start <= 23) {
+            for (int i = 0; i < 24; ++i) {
+                validHours.emplaceBack(sHours.start);
+            }
+        }
+        if (sHours.end > 1) {
+            for (int i = 0; i < 23; ++i) {
+                validHours.emplaceBack(i);
+            }
+        }
+        auto rand = sead::Random();
+        validHours.shuffle(&rand);
+        mFallHour = *validHours[0];
+    } else {
+        mFallHour = sead::GlobalRandom::instance()->getS32Range(sHours.start, sHours.end - 1);
+    }
+    mFallMinute = sead::GlobalRandom::instance()->getU32(59);
+    mInitialised = true;
+}
+
+void ShootingStarMgr::calc_() {
+    if (Manager::instance() != nullptr) {
+        if (!mIsMainField)
+            return;
+        if (sScheduled) {
+            if (Manager::instance()->isDemo()) {
+                sScheduled = false;
+                return;
+            }
+            TimeMgr* tm = Manager::instance()->getTimeMgr();
+            if (tm != nullptr) {
+                int hour = tm->getHour();
+                if (mInitialised) {
+                    if (hour == mFallHour && mFallMinute < tm->getMinute())
+                        mNeedSpawnStar = true;
+                } else if (Manager::instance() != nullptr) {
+                    if (isScheduledTime())
+                        initSchedule();
+                }
+            }
+        }
+    }
+    if (mNeedSpawnStar && checkCamera()) {
+        spawnShootingStar();
+        mInitialised = false;
+        mNeedSpawnStar = false;
+    }
+}
+
+bool ShootingStarMgr::isScheduledTime() {
+    bool result = false;
+    if (Manager::instance() != nullptr) {
+        TimeMgr* tm = Manager::instance()->getTimeMgr();
+        if (tm != nullptr) {
+            int hour = tm->getHour();
+            if (sHours.start > sHours.unk) {
+                result = sHours.unk < hour && hour < sHours.start;
+            } else {
+                result = sHours.unk < hour || hour < sHours.start;
+            }
+        }
+    }
+    return result;
+}
+
+void ShootingStarMgr::resetStarPosition() {
+    if (gdt::Manager::instance()) {
+        gdt::Manager::instance()->setVec3f(sead::Vector3f::zero,
+                                           sead::SafeString("ShootingStarPosition"));
```

Comment:
```suggestion
                                           "ShootingStarPosition");
```
sead::SafeString can be implicitly constructed from a `const char*` so explicitly writing sead::SafeString isn't necessary (and it's somewhat verbose)